### ISSUE TITLE
Polish Pass: Utility Features

### ISSUE THREAD

**Issue by endigo9740** [skeletonlabs/skeleton#922]
> **Warning** - this ticket is a work in progress. If you have suggestions please add a comment below!

Base on:
- https://github.com/skeletonlabs/skeleton/issues/782

## Goals

The goal of this pass will be to improve, refine, and finalize all Utility features in Skeleton.

- [x] Fix and address known issues and bugs
- [x] Refactor code, cleanup styles, and improve edge cases
- [x] Improve documentation for each feature
- [x] Expand with new features as relevant
- [x] Make overlays not suck, or at least suck less
- [x] Ensure Vitest test cases are passing
- [x] Resolve `FIXME` issues ([reference](https://user-images.githubusercontent.com/1509726/216708976-ea8c37a1-adbf-46e5-a09f-095402907c6e.png))
- [x] Update all style props to use `CssClasses` type

### Local Storage Stores

- [x] Update to the latest upstream version 0.40.0

### Code Block

- [x] Minor visual refresh and code refactor
- [x] Property `buttonCopy` renamed `button`, default value is now `variant-soft`
- [x] Added `buttonLabel` and `buttonCopied` for basic i18n support
- [x] Added style props for `blur` and `shadow`
- [x] Prepare documentation "previewer" component

### Drawers

- [x] Docs: make it clear these are a singleton
- [x] Greatly improved documentation and setup instructions
- [x] Provide advanced tip for animating background contents, per [#637](https://github.com/skeletonlabs/skeleton/issues/637)
- [x] Refactor of Drawer animation logic
- [x] Added `shadow` property for Drawer element
- [x] Added `regionBackdrop|Drawer` props. Available as overrides as well!
- [x] BREAKING: Removed drawer `margin` in favor of backdrop `padding` (provides better results)
- [x] Rounded corners by default, based on theme settings (set `rounded-none` to disable)

### Lightswitch

> **Warning** - this utility feature has been complete rebuilt from the ground up.

- [x] Lightswitch component rebuilt with new UI design and new props available
- [x] Now uses a service-like approach, per [#737](https://github.com/skeletonlabs/skeleton/issues/737)
- [x] Provided an alternative to to the component if you wish to use OS preference automatically
- [x] Users can now interface with our methods, or build their own Lightswitch UI
- [x] Provided and documented interface stores, getters, and setter methods.

> NOTE: we need to test [#905](https://github.com/skeletonlabs/skeleton/issues/905)

### Modals

- [x] Docs: make it clear these are a singleton
- [x] Implement system to register and reuse custom modals, per [#720](https://github.com/skeletonlabs/skeleton/issues/720)
- [x] Custom modals should be "naked" by default and allow FULL customization
- [x] Allow prompt input to accept custom validation attributes, per [#899](https://github.com/skeletonlabs/skeleton/issues/899)
- [x] Added image modal example
- [x] Move modal example components to `/docs` - should not be part of the public package
- [x] Provide prop to set position - ex: top (search) or center (default), ([reference](https://user-images.githubusercontent.com/31982384/214623340-22252a12-fe33-42c8-9906-c40d385df4b8.png))
- [x] Improve component scrolling, per [#890](https://github.com/skeletonlabs/skeleton/issues/890)

> NOTE: given the potential scale and scope of changes required, we'll have to circle back to [#937](https://github.com/skeletonlabs/skeleton/issues/937)

### Toasts

- [x] Docs: make it clear these are a singleton
- [x] Wrapping button element needs selector like `.toast-actions`
- [x] Update button style props to allow changing button type
- [x] Added new `buttonDismissLabel` prop to override the dismiss button label text
- [x] Replaced `preset` prop with `background` property allowing any variant or background class.
- [x] Improved animation using Svelte FLIP 
- [x] Added a `callback` function feature for the ToastSettings object

### Popups

> **Warning** - this completely replaces Menus and Tooltip utilities
> Requires new dependency: https://floating-ui.com/

- [x] Replaced Menus and Tooltips with new "Popup" utility feature
- [x] Fully documented, has tons of config options
- [x] Updated examples, including a functional Combobox
- [x] Includes a fade animation on open/close
- [x] Supports multiple interaction event handlers (click, hover, both)
- [x] Can optionally include a customizable arrow on any variation
- [x] Now includes 12 position options
- [x] Dramatically better performance in keeping the popup on screen
- [x] Fully tree shakable
- [x] Allows you to modify a host of settings through "middleware"
- [x] Supports a11y features, such as keyboard interaction

### Menus

> DEPRECATED: replaced with Popups (shown above)

### Menus

> DEPRECATED: replaced with Popups (shown above)

### Data Tables

> See our our audit of prosed third party solutions here [#821](https://github.com/skeletonlabs/skeleton/issues/821)

**Comment by endigo9740** [comment]
A new PR has been update, and you can preview the changes here:
https://skeleton-docs-git-chore-utilities-polish-pass-skeleton-labs.vercel.app/utilities/codeblocks

**Comment by endigo9740** [comment]
Today's updates have focused on Drawers, including a refactor to cleanup the animation logic, and a slew of new style properties. This should greatly expand the customization capabilities of this feature. Here's an example of a highly customized instance:

![Screenshot 2023-02-08 at 4 57 02 PM](https://user-images.githubusercontent.com/1509726/217670299-eea06db4-964a-49b5-80e8-5f31ffbd41b6.png)

Additionally I've research a means to animate the background content while the drawer is open. This will be best served as a feature users implement, as we cannot know what sort of layout structure they might be using.

![Screenshot 2023-02-08 at 4 58 07 PM](https://user-images.githubusercontent.com/1509726/217670466-def7a9b8-9e4e-4aac-9c64-6146968509e9.png)

Additionally, more information the Singleton pattern this uses has been added, as well as a note that we do not (and will not) support a means to disable the click-to-close backdrop feature. It's something we get regular requests for, so I want to make it clear it's against a11y best practices.

**Comment by endigo9740** [comment]
The Lightswitch component has been completely rebuilt from the ground up. Preview it here:
https://skeleton-docs-git-chore-utilities-polish-pass-skeleton-labs.vercel.app/utilities/lightswitches

![Screenshot 2023-02-12 at 3 57 01 PM](https://user-images.githubusercontent.com/1509726/218339695-21d43147-38ef-4c90-9baf-30139748a383.png)

Notable changes:

- New UI design that adheres to theme settings
- A variety of new props, including a means to adjust the size
- Underlaying logic completely rebuilt, abstracted to an external service
- This service exposes the stores, getters, and setting methods for users
- Users can utilize these to generate their own Lightswitch UI component
- Also provided an alternative if you wish to use OS preference only

In a nutshell, the basic version is at simple as before, but now provide a ton of power-user features.

**Comment by endigo9740** [comment]
We have a BIG series of modal updates today. These can be previewed here:
https://skeleton-docs-git-chore-utilities-polish-pass-skeleton-labs.vercel.app/utilities/modals

![Screenshot 2023-02-13 at 5 06 57 PM](https://user-images.githubusercontent.com/1509726/218594688-8bcdc50d-c0dd-44b8-9c81-0c71a7c98d02.png)

Notable changes:

- You can now optionally register components on the parent `<Modal>` for easy reuse. 
- Custom modals are no longer styled by default, giving you complete control over design
- ModalSettings now allows for `valueAttr` to define the prompt input type and other attributes
- I've added a new example modal, which displays a single image with a detached close button.
- You can now easily set the modal vertical position either globally or per instance.
- The backdrop region now support vertical scroll for tall components.

That said, I do still have one more item I would like to address, which is making modals async by default, per [#937](https://github.com/skeletonlabs/skeleton/issues/937)

**Comment by endigo9740** [comment]
Several notable updates to the Toast utility today. You can preview this here:
https://skeleton-docs-git-chore-utilities-polish-pass-skeleton-labs.vercel.app/utilities/toasts

![Screenshot 2023-02-14 at 1 41 41 PM](https://user-images.githubusercontent.com/1509726/218841417-abf3fcf2-f172-441a-ae4e-7c4656fc6089.png)

- Minor UI adjustments
- Various new or updates prop values, including the ability to customize the dismiss button text
- Replaced the `preset` prop with a `background` prop which can take utility classes or variants
- Improved the animation, especially when multiple are queued

**Comment by Hugos68** [comment]
All visible toasts will slighly bump up a tiny amount after every toast that is triggered when the visible limit of 3 is full.

**Comment by endigo9740** [comment]
Big update today - both Menus and Tooltips have now been replaced with the new "Popups" feature.

https://user-images.githubusercontent.com/1509726/219174651-1f5749ec-9810-4e72-8512-36d354988dae.mp4

This includes a new dependency called Floating UI (previously known as Popper.js):
https://floating-ui.com/

Notable improvements:

- Fully documented, has tons of config options
- Updated examples, including a functional Combobox
- Includes a fade animation on open/close
- Supports multiple interaction event handlers (click, hover, both)
- Can optionally include a customizable arrow on any variation
- Now includes 12 position options
- Dramatically better performance in keeping the popup on screen
- Fully tree shakable
- Allows you to modify a host of settings through "middleware"

**Comment by ryceg** [comment]
Minor quibble: you use the error token for the "advanced" tag, and should be using the error token.

**Comment by niktek** [comment]
- combobox opens once on iOS and not at all on Android

**Comment by AdrianGonz97** [comment]
feat: Share the open/close state of the popup, similar to how we currently do it for menus.

**Comment by endigo9740** [comment]
NOTE: mention that Floating UI popups have a fighting z-index issue when overlaying elements using `backdrop-blur` (at least on Chrome). It's most likely a browser bug.

**Comment by endigo9740** [comment]
Reminder:

- [ ] Implement @ryceg QA feedback items in the PR thread: https://github.com/skeletonlabs/skeleton/pull/936

**Comment by JustBarnt** [comment]
- [ ] Toast - Implement functionality below to allow Toasts to trigger a callback when it is created and closed.

- So when the ID is assigned -> trigger the callback with id: whatever and status: assigned as the response
- When the toast is closed -> we trigger it again, but this time change status: closed

The callback function would then look for a status of "assigned" or "closed" to fire the appropriate event.

**Comment by endigo9740** [comment]
FYI this has been merged into `dev` branch, please continue QA testing there!
https://dev.skeleton.dev/

**Comment by SebasF1349** [comment]
- [ ] Code Blocks: First example (at the very top) shows `<div>This is meta</div>` in the code but `<p>Hello Skeleton<p>` in the preview.
- [ ] Code Blocks: Same examples, in the code view, have incorrect colors. Probably the template literals in code should be blue, but some chars (`<`, `{`, `=`) break it?
![brave_lWzecyDUBG](https://user-images.githubusercontent.com/33849340/219720569-e06cd6db-91cf-4632-80ab-cace798c8737.png)
- [ ] Toasts: Don't know why or if it's intended, but the Basic example logs id and status.
- [ ] Toasts: After spamming toasts and closing them I keep getting this error (it works btw)
![brave_LFG9IUOmWN](https://user-images.githubusercontent.com/33849340/219731688-97d6cb0e-4265-4466-a2fc-3cc6713d1195.png)

**Comment by endigo9740** [comment]
Per Popups

- [ ] We need to implement the `interactive` mode option to allow popups to close when interaction or route change occurs.

**Comment by SebasF1349** [comment]
- [ ] Modals: `undefined` class
![msedgewebview2_xNkQxBRusk](https://user-images.githubusercontent.com/33849340/219772961-a89a01ba-7e90-475c-9b05-a49a74dc8854.png)

**Comment by endigo9740** [comment]
I've implement a new `queryClose` setting for popups that allow you to set the list of the popup's child elements that will close the popup when clicked. By default any anchor or button clicked within the popup will automatically close it, but this can be easily overrided.
